### PR TITLE
Add AG Grid WASM Bindings and Component Coverage #2

### LINKED ISSUE #2: BOUNTY: Coverage for widget library bindings and options
We are using several open source libraries to provide widget components. The WASM bindings and options are very incomplete and coverage needs to be improved. 

Coverage of libraries should at least include: 

- All initial option object configurations
- All event types
- Demo of Yew functional components showcasing Nostr integration

### PR DESCRIPTION
**Pull request by rahat2134**
Implemented AG Grid WASM bindings and Yew component integration with comprehensive coverage for simple table structures.
(Fix #2  - AG grid)
### Features Added
- WASM bindings for AG Grid core functionality
- Column definitions and configuration
- Grid options (pagination, row selection)
- Reusable Yew component
- Auto-sizing and refresh capabilities

### Implementation Details
1. AG Grid Bindings (`ag_grid.rs`):
   - Core grid creation and manipulation
   - Column definitions with customization options
   - Default column properties
   - Grid options configuration
   - Type-safe option builders

2. Yew Component (`component.rs`):
   - Reusable functional component
   - Props for data, columns, and configuration
   - Auto grid initialization and updates
   - Responsive layout support

### Files Changed
- `src/widgets/ag_grid/ag_grid.rs`
- `src/widgets/ag_grid/component.rs`
- `src/widgets/ag_grid/mod.rs`

### Features
- Simple table structures
- Initial option configurations
- Event handling
- Nostr integration ready

Closes #2

### CODE REVIEW

**Comment by rahat2134** [comment]
@42Pupusas Please take a look!

**Comment by rahat2134** [comment]
@42Pupusas  PTAL to this after this I will tackle others on the list...
For e.g., leaflets and calendar

**Comment by 42Pupusas** [comment]
Hey @rahat2134  this bindings and helpers are looking great. 

Please note the bounty description in #2 includes: 

```
Demo of Yew functional components showcasing Nostr integration
```

A simple test component that puts kind 0 or 1 events in a table will suffice.

You can see an example of pulling notes from the relay pool in the code I committed today.

**Comment by rahat2134** [comment]
@42Pupusas PTAL! I have pushed the changes.

File: src/widgets/ag_grid/test_component.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            kind: note.get_kind(),
+        }
+    }
+}
+
+#[function_component(NostrNotesGrid)]
+pub fn nostr_notes_grid() -> Html {
+    let relay_ctx = use_context::<NostrProps>().expect("No relay context found");
+    
+    // Convert notes to row data
```
Comment (42Pupusas) [inline src/widgets/ag_grid/test_component.rs]: The update of the notes to rows needs to happen in a `use_effect` or the thread will block and not update.

```
    yew::use_effect_with(relay_ctx.unique_notes.clone(), move |notes| {
         // Update the rows here
        || {
             // Cleanup logic here
        }
    });
```

**Comment by 42Pupusas** [comment]
Im  getting a browser panic when the table is created, can we make sure all the bindings are implemented correctly?

```
ag-grid-community.min.js:1 Uncaught TypeError: e.map is not a function
    at u.setRowData (ag-grid-community.min.js:1:1713)
    at p.setRowData (ag-grid-community.min.js:1:20629)
    at y.onRowDataUpdated (ag-grid-community.min.js:1:28291)
    at ag-grid-community.min.js:1:27064
    at ag-grid-community.min.js:1:43663
    at kg.wrapIncoming (ag-grid-community.min.js:1:669896)
    at ag-grid-community.min.js:1:43645
    at ag-grid-community.min.js:1:43703
    at Set.forEach (<anonymous>)
    at r (ag-grid-community.min.js:1:43552)

﻿```

File: src/widgets/ag_grid/test_component.rs (hunk 1 above)
Comment (42Pupusas) [reply src/widgets/ag_grid/test_component.rs]: Notice the pool does not refresh the events when the table is rendered as it is blocking the thread.  

image


Without table:

image

**Comment by rahat2134** [comment]
@42Pupusas PTAL! Any more suggestions?

**Review by 42Pupusas — CHANGES_REQUESTED** [review]
Please check this comments @rahat2134

**Comment by rahat2134** [comment]
@42Pupusas , Pushed the changes. PTAL!
Screenshot 2024-10-28 at 2 56 58 AM

**Review by 42Pupusas — APPROVED** [review]

**Comment by 42Pupusas** [comment]
Everything looks and works great. Another great contribution @rahat2134 .

**Comment by rahat2134** [comment]
@42Pupusas, Is the transaction for this bounty completed?

**Comment by rahat2134** [comment]
@42Pupusas PTAL whenever you have some bandwidth.

**Pull request merged**